crimson/os/seastore: add multi-device support

xxhdx1985126 opened the pull request:
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

athanatos commented on src/crimson/os/seastore/scanner.h:
```diff
@@ -10,17 +10,13 @@
+class TransactionManager;
 
-class Scanner {
+class Scanner : public ExtentReader {
```
This relationship doesn't make sense.  Why not simply rename Scanner to ExtentReader and remove the ExtentReader interface from SegmentManager?  I don't think multiple ExtentReader implementations is desirable.

xxhdx1985126 replied:
will change it, I just thought both SegmentManager and Scanner have to same `read()` interface, so I made tho ExtentReader an interface

athanatos replied:
SegmentManager is an interface because we expect to have different implementation for different physical device interfaces (ZNS, conventional block, some kind of pmem, etc).  ExtentReader is pretty much just a mapping from device_id_t to the different SegmentManager backend implementations, so it doesn't seem like it'll ever have another implementation.  Removing the interface and clarifying the type and variable names will make the intent clearer to readers.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,29 +36,139 @@ struct seastore_meta_t {
 };
 
 // Identifies segment location on disk, see SegmentManager,
```
This comment referred to segment_id_t.  Please add a comment like this for every type you introduce below or elsewhere in this PR.

xxhdx1985126 replied:
will do

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,29 +36,139 @@ struct seastore_meta_t {
+  0xF << (std::numeric_limits<device_segment_id_t>::digits - device_id_len);
+
+struct segment_id_t {
+  device_segment_id_t segment;
```
Please audit this PR for uninitialized member variables like this one.  device_segment_id_t = internal_segment_id_t = uint32_t which does not have a default initializer, so your default constructor below will leave it uninitialized.

xxhdx1985126 replied:
I added internal_segment_id_t to represent a segment id without device id, I thought this might be useful when we are referencing segment ids within the single segment manager scope. Am I right?

athanatos replied:
This comment referred specifically to the fact that it will have undefined contents after the default constructor is called.  Change the default constructor to set itself to NULL_SEG_ID.  There aren't really any cases where it's ok to leave a variable uninitialized.

xxhdx1985126 replied:
Yes, sorry that I left the comment at the wrong place, it should have been added [here](https://github.com/ceph/ceph/pull/43181#discussion_r712669743)

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,29 +36,139 @@ struct seastore_meta_t {
+  }
+
+  DENC(segment_id_t, v, p) {
+    DENC_START(1, 1, p);
```
We don't want to version segment_id_t, just call denc(v.segment, p).

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,29 +36,139 @@ struct seastore_meta_t {
 
 // Identifies segment location on disk, see SegmentManager,
-using segment_id_t = uint32_t;
+using internal_segment_id_t = uint32_t;
```
I see internal_segment_id_t and device_segment_id_t used almost interchangeably in this PR.  What does each one mean?  Where is each meant to be used?

xxhdx1985126 replied:
I added internal_segment_id_t to represent a segment id without device id, I thought this might be useful when we are referencing segment ids within the segment manager scope. Am I right?

athanatos replied:
Ok, what does device_segment_id_t mean?

xxhdx1985126 replied:
segment ids with a device id encapsulated, it's used in segment_id_t. I think I see your point, maybe I should hide device_segment_id_t from outside segment_id_t, am I right?

athanatos replied:
Right, only one of those two types should be visible outside of segment_id_t.  I'd also like you to swap them.  internal_segment_id_it should be "internal" to segment_id_t and not visible outside of it (you can define the type in the private: section of segment_id_t).  Let device_segment_id_t be the one used externally.

xxhdx1985126 replied:
Got it:-)

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -44,6 +89,199 @@ struct segment_info_t {
+// this segment_info_set_t class is intend to avoid iterating a device map
+// for a specific device, which could be expensive if we do it for every io
+template <typename SegmentInfoT>
+class segment_info_set_t {
```
It seems like this could just use a std::map<device_id_t, std::vector<std::optional<SegmentInfoT>> internally.  I don't see  a justification for the complexity here.  If this is necessary, I'd expect the comment above to specifically explain why a simpler structure won't work as well as a concise and complete explanation of the internal invariants.

xxhdx1985126 replied:
Actually, I thought about using std::map<device_id_t, std::vector<std::optional>>. However, I thought if we use the top 4 bits of segment id as the device id, which means we may have up to 16 devices, and `segment_info_set_t::operator[]` is called for every journal writes when we do `SegmentCleaner::mark_space_used` or `SegmentCleaner::mark_space_free`, so I was worried that doing a 16-element search for every transaction is a bit expensive. Honestly, I was not sure if I was considering this right.

athanatos replied:
Without evidence that it actually makes a difference, this complexity isn't worthwhile.  Please swap to the simpler implementation.

athanatos replied:
Also, std::vector<std::vector<std::optional>> would still be simpler.

xxhdx1985126 replied:
Got it, will do:-)

xxhdx1985126 replied:
@athanatos Um, it seems that except for the perf reason, we also need a place to keep each segment manager's empty segments, opened segments and other similar stuff so that we can keep track of things like each segment manager's space usage. So I think maybe it's worthwhile to keep segment_info_set_t as a structure to hold that information. I'm changing segment_info_set_t as follows:

```
struct segment_manager_info_t {
    std::vector<SegmentInfoT> segment_infos
};

struct segment_info_set_t {
    std::vector<std::optional<segment_manager_info_t>> sm_info_vec;
};
```

What do you think? Thanks.

athanatos replied:
If I'm understanding that correctly, that seems reasonable.  Primarily, I find the segment_offset indirection unnecessarily complex.  As long we avoid that by embedding a seperate segment_infos vector for each manager, it should be simple enough.

xxhdx1985126 replied:
Got it:-)

athanatos commented on src/crimson/os/seastore/transaction_manager.h:
```diff
@@ -491,17 +492,32 @@ class TransactionManager : public SegmentCleaner::ExtentCallbackInterface {
   friend class Transaction;
 
-  SegmentManager &segment_manager;
+  SegmentManager &segment_manager; // primary device
```
I don't see an explanation of what "primary device" means.  Does it mean that we only journal to this segment_manager?

xxhdx1985126 replied:
yes, that's it

athanatos commented:
Seems like a good start!  I think the first few things to address are probably:
1) Try to simplify segment_info_set_t as much as possible using existing stl types if at all possible.
2) Clarify usage of internal_segment_id_t vs device_segment_id_t -- I think you probably want to eliminate usage of internal_segment_id_t except within the definition of segment_id_t and use device_segment_id_t everywhere else.
3) Add an explanation of what "primary device" means.  Probably deserves to be put in the comment at the top of TransactionManager.
4) Scanner->ExtentReader -- make sure to update variable names as appropriate as well.  This should probably be the first commit in the PR including the updates to Cache.

cyx1231st commented on src/crimson/os/seastore/seastore.h:
```diff
@@ -263,7 +264,10 @@ class SeaStore final : public FuturizedStore {
     OMapManager::omap_list_config_t config);
 
+  std::string root;
+  seastar::file root_f;
```
Why not move this to `SeaStore::mkfs()` as it is the only user?

xxhdx1985126 commented:
@athanatos @cyx1231st I've addressed all the issues, and splitted the original commits into smaller ones, please take a look again, thanks:-)

cyx1231st commented on src/crimson/os/seastore/segment_manager/block.cc:
```diff
@@ -533,61 +558,75 @@ SegmentManager::read_ertr::future<> BlockSegmentManager::read(
 {
+  logger().debug("{} {}", __func__, get_device_id());
   namespace sm = seastar::metrics;
   // TODO: add label for device_id
```
nit, remove the implemented TODO.

xxhdx1985126 replied:
done:-)

cyx1231st commented on src/crimson/os/seastore/segment_manager/block.cc:
```diff
@@ -533,61 +558,75 @@ SegmentManager::read_ertr::future<> BlockSegmentManager::read(
   stats.reset();
   metrics.add_group(
-    "segment_manager",
+    std::string("segment_manager"),
```
nit, drop the unnecessary change.

xxhdx1985126 replied:
done:-)

liu-chunmei commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
 // Identifies segment location on disk, see SegmentManager,
-using segment_id_t = uint32_t;
+struct segment_id_t {
+  internal_segment_id_t segment = DEFAULT_INTERNAL_SEG_ID;
```
does the internal_segment_id_t always include device_id_t in the high 4 bits?

liu-chunmei replied:
what is the benefit to use a combined integer to save device_id_t and device_segment_id_t, every place when to use them need calculate.  how about use a structure with two members?

athanatos replied:
This is fine.  Otherwise, you'd need to use a bitfield.

liu-chunmei commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+  }
+
+  [[gnu::always_inline]]
+  internal_segment_id_t internal_segment_id() const {
```
should use device_segment_id_t to be clearly understand even though they are same type.

athanatos replied:
Correct, literally nothing outside of this class should see internal_segment_id_t.

athanatos replied:
Rename this to get_device_segment_id() and have it return device_segment_id_t as suggested by @liu-chunmei elsewhere.

xxhdx1985126 replied:
will do

liu-chunmei commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+  segment_id_t() = default;
+  constexpr segment_id_t(internal_segment_id_t segment)
+    : segment(segment) {}
+  segment_id_t(device_id_t id, internal_segment_id_t segment)
```
the second one should be device_segment_id_t.

liu-chunmei commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+    segment_id_t old = *this;
+    --segment;
+    return old;
+  }
```
what does this do? how do you deal with input int?

liu-chunmei commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+    segment_id_t old = *this;
+    ++segment;
+    return old;
+  }
```
what does this do? how do you deal with input int?

athanatos replied:
operator++(int) in C++ denotes the postfix operator (i++). operator++() in C++ denotes the prefix operator (++i).  The int aram is meaningless except to distinguish which you are defining.  See https://en.cppreference.com/w/cpp/language/operator_incdec

liu-chunmei replied:
Thanks, understand.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/block.cc:
```diff
@@ -343,8 +343,10 @@ Segment::write_ertr::future<> BlockSegment::write(
     segment_id_t id, segment_off_t write_pointer)
 {
+  assert(id.device_id() == superblock.device_id);
+  auto s_id = id.internal_segment_id();
```
this get device_segment_id_t, right?  how about rename it device_segment_id()?

liu-chunmei commented on src/crimson/os/seastore/segment_manager/block.cc:
```diff
@@ -446,26 +449,29 @@ BlockSegmentManager::mkfs_ret BlockSegmentManager::mkfs(seastore_meta_t meta)
 {
-  if (id >= get_num_segments()) {
+  assert(id.device_id() == superblock.device_id);
+  auto s_id = id.internal_segment_id();
```
same here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/block.cc:
```diff
@@ -479,24 +485,26 @@ SegmentManager::open_ertr::future<SegmentRef> BlockSegmentManager::open(
   segment_id_t id)
 {
+  assert(id.device_id() == superblock.device_id);
+  auto s_id = id.internal_segment_id();
```
same here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/block.cc:
```diff
@@ -507,7 +515,8 @@ SegmentManager::read_ertr::future<> BlockSegmentManager::read(
 {
-  if (addr.segment >= get_num_segments()) {
+  assert(addr.segment.device_id() == get_device_id());
+  if (addr.segment.internal_segment_id() >= get_num_segments()) {
```
same here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/block.cc:
```diff
@@ -522,11 +531,11 @@ SegmentManager::read_ertr::future<> BlockSegmentManager::read(
   }
 
-  if (tracker->get(addr.segment) == segment_state_t::EMPTY) {
+  if (tracker->get(addr.segment.internal_segment_id()) == segment_state_t::EMPTY) {
```
also here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/block.cc:
```diff
@@ -522,11 +531,11 @@ SegmentManager::read_ertr::future<> BlockSegmentManager::read(
       "BlockSegmentManager::read: read on invalid segment {} state {}",
       addr.segment,
-      tracker->get(addr.segment));
+      tracker->get(addr.segment.internal_segment_id()));
```
also here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/block.h:
```diff
@@ -213,9 +221,11 @@ class BlockSegmentManager final : public SegmentManager {
   size_t get_offset(paddr_t addr) {
     return superblock.first_segment_offset +
-      (addr.segment * superblock.segment_size) +
+      (addr.segment.internal_segment_id() * superblock.segment_size) +
```
here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/ephemeral.cc:
```diff
@@ -59,10 +59,11 @@ Segment::write_ertr::future<> EphemeralSegment::write(
 Segment::close_ertr::future<> EphemeralSegmentManager::segment_close(segment_id_t id)
 {
-  if (segment_state[id] != segment_state_t::OPEN)
+  auto s_id = id.internal_segment_id();
```
same here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/ephemeral.cc:
```diff
@@ -80,7 +81,7 @@ Segment::write_ertr::future<> EphemeralSegmentManager::segment_write(
     bl.length(),
     bl.crc32c(1));
-  if (!ignore_check && segment_state[addr.segment] != segment_state_t::OPEN)
+  if (!ignore_check && segment_state[addr.segment.internal_segment_id()] != segment_state_t::OPEN)
```
same here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/ephemeral.cc:
```diff
@@ -145,41 +146,43 @@ void EphemeralSegmentManager::remount()
   segment_id_t id)
 {
-  if (id >= get_num_segments()) {
+  auto s_id = id.internal_segment_id();
```
same here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/ephemeral.cc:
```diff
@@ -145,41 +146,43 @@ void EphemeralSegmentManager::remount()
 SegmentManager::release_ertr::future<> EphemeralSegmentManager::release(
   segment_id_t id)
 {
+  auto s_id = id.internal_segment_id();
```
same here.

liu-chunmei commented on src/crimson/os/seastore/segment_manager/ephemeral.h:
```diff
@@ -56,7 +56,7 @@ class EphemeralSegmentManager final : public SegmentManager {
 
   size_t get_offset(paddr_t addr) {
-    return (addr.segment * config.segment_size) + addr.offset;
+    return (addr.segment.internal_segment_id() * config.segment_size) + addr.offset;
```
same here.

liu-chunmei commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -379,14 +650,12 @@ class SegmentCleaner : public SegmentProvider {
 
   SpaceTrackerIRef space_tracker;
-  std::vector<segment_info_t> segments;
+  segment_info_set_t<segment_info_t> segments;
```
can we rename segments to seg_infos  or segment_infos, always get confused when code has segment variable which stand for segment_id_t.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+// this is a unsigned int with the top "device_id_len"
+// bits representing the id of the device on which the
+// segment resides
+using internal_segment_id_t = uint32_t;
```
Define this type as private within segment_id_t.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+// bits representing the id of the device on which the
+// segment resides
+using internal_segment_id_t = uint32_t;
+// type of device id
```
Replace with 'identifier used to distinguish between physical devices within a seastore instance'

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+  bool operator>=(const segment_id_t& other) const {
+    return segment >= other.segment;
+  }
+  segment_id_t& operator++() {
```
Are there any users that actually increment and decrement this type without converting to device_segment_id_t first?  Should we even allow that?

xxhdx1985126 replied:
src/test/crimson/seastore/test_seastore_cache.cc:49 and src/test/crimson/seastore/test_seastore_journal.cc:91 are using these operators

athanatos replied:
I don't think adding an integer type to segment_id_t has an obvious meaning (to a reader, it could increment the device, the segment, or both if at the end of a device).  This implementation in particular can cause the segment_id_t to no longer refer to an actual segment.  No user outside of SegmentManager implementations (which will be using device_segment_id_t) or special cases (like these tests where we are basically emulating a SegmentManager) should be manipulating the segment value, so exposing the ability to increment it through the interface invites misuse.

I think you should make those two users (as well as test_btree_lba_manager.cc:59) explicitly construct new segment_id_t instances and remove these operators.

```
--- a/src/test/crimson/seastore/test_btree_lba_manager.cc
+++ b/src/test/crimson/seastore/test_btree_lba_manager.cc
@@ -51,9 +51,14 @@ struct btree_lba_manager_test :
 
   segment_id_t next = 0;
   get_segment_ret get_segment(device_id_t id = 0) final {
+    auto ret = next;
+    next = segment_id_t{
+      next.device_id(),
+      next.internal_segment_id() + 1
+    };
     return get_segment_ret(
       get_segment_ertr::ready_future_marker{},
-      next++);
+      ret);
   }
 
   void set_segment_written_to(paddr_t addr) final {}
diff --git a/src/test/crimson/seastore/test_seastore_cache.cc b/src/test/crimson/seastore/test_seastore_cache.cc
index 1b131869fa8..ceef5a0ae2b 100644
--- a/src/test/crimson/seastore/test_seastore_cache.cc
+++ b/src/test/crimson/seastore/test_seastore_cache.cc
@@ -46,7 +46,11 @@ struct cache_test_t : public seastar_test_suite_t {
                segment_manager->get_segment_size());
     if (current.offset + (segment_off_t)bl.length() >
        segment_manager->get_segment_size())
-      current = paddr_t{current.segment + 1, 0};
+      current = paddr_t{
+       segment_id_t{
+         current.segment.device_id(),
+         current.segment.internal_segment_id() + 1
+       }, 0};
 
     auto prev = current;
     current.offset += bl.length();
```

xxhdx1985126 replied:
will do:-)

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+    --segment;
+    return old;
+  }
+  segment_id_t operator+(internal_segment_id_t t) {
```
t is coming from the outside, *must* be device_segment_id_t.

athanatos replied:
Audit every other user of this type as well.

xxhdx1985126 replied:
> t is coming from the outside, _must_ be device_segment_id_t.

will correct it.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+
+// mask for segment manager id
+constexpr internal_segment_id_t SM_ID_MASK =
+  0xF << (std::numeric_limits<internal_segment_id_t>::digits - device_id_len);
```
Implementation detail of segment_id_t, move to the private: area of segment_id_t.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+  0xF << (std::numeric_limits<internal_segment_id_t>::digits - device_id_len);
+// default internal segment id
+constexpr internal_segment_id_t DEFAULT_INTERNAL_SEG_ID =
+  std::numeric_limits<internal_segment_id_t>::max() - 1;
```
Implementation detail of segment_id_t, move to the private: area of segment_id_t.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
+constexpr uint16_t max_devices = 1 << device_id_len;
+
+// segment ids without a device id encapsulated
+using device_segment_id_t = internal_segment_id_t;
```
Because internal_segment_id_t is defined within segment_id_t, just declare this as

using device_segment_id_t = uint32_t.

athanatos commented on src/test/crimson/seastore/test_seastore_journal.cc:
```diff
@@ -117,8 +121,9 @@ struct journal_test_t : seastar_test_suite_t, SegmentProvider {
 	return crimson::do_for_each(
-	  boost::make_counting_iterator(segment_id_t{0}),
-	  boost::make_counting_iterator(segment_manager->get_num_segments()),
+	  boost::make_counting_iterator(internal_segment_id_t{0}),
```
device_segment_id_t

athanatos commented on src/crimson/os/seastore/cache.cc:
```diff
@@ -20,8 +20,11 @@ using std::string_view;
-  segment_manager(segment_manager)
+Cache::Cache(
+  ExtentReader &reader,
+  SegmentManager& segment_manager)
```
I don't think we need a segment_manager instance here.  The only remaining user is prepare_record, which calls get_block_size().  I think block size needs to be global throughout seastore and will be set at mkfs time.  Can you add it to the superblock and pass it in explicitly to Cache instead of the SegmentManager instance?

athanatos replied:
For now, you can just pass in the block_size to Cache.  We can worry about putting the block size into the superblock later.

athanatos commented:
@xuxuehan Once you've addressed the comments I've left so far, I'd like to merge the first three commits on their own and then continue reviewing the rest of the commits.  The segment_id_t and ExtentReader changes will probably be ready to merge after those comments are addressed and are likely to cause conflicts with other ongoing work, so merging them quickly is worthwhile.

I'll continue reviewing the remaining commits so that we can merge them pretty quickly after that.  Does that seem reasonable?

athanatos commented on src/crimson/os/seastore/extent_placement_manager.cc:
```diff
@@ -273,7 +277,9 @@ SegmentedAllocator::Writer::roll_segment(bool set_rolling) {
 
-  return segment_provider.get_segment().safe_then([this](auto segment) {
+  return segment_provider.get_segment(
+    segment_manager.get_device_id()
```
Does this allocator actually need a SegmentManager instance?  Can we just give it a device id?

xxhdx1985126 replied:
Um, segment manager is also used to open/close segments in the writer

athanatos replied:
Ah, right.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -18,15 +18,66 @@
+using unavail_segment_info_set_t =
+  std::list<segment_info_t*>;
+using unavail_segment_info_iterator =
+  unavail_segment_info_set_t::iterator;
```
Move from here to within segment_manager_info_t

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -18,15 +18,66 @@
+  {}
+
+  void init_segment_infos(SegmentInfoT&&);
+
```
```
using info_set_t = std::list<segment_info_t*>;
using iterator = info_set_t::iterator
```

Please make a point of scoping type declarations for a struct within that struct.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -18,15 +18,66 @@
+  // we need to keep track of opened segments so that
+  // we can involve the used/free space into seastore
+  // disk space calculation
+  unavail_segment_info_iterator usii;
```
segment_manager_info_t::iterator

athanatos replied:
This should go away once the unavailable segment handling is simplified.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -18,15 +18,66 @@
+
+  void init_segment_infos(SegmentInfoT&&);
+
+  unavail_segment_info_set_t usis;
```
This is a secondary index of the contents of segment_infos containing those segments which are unavailable.  You really need to include a comment with an explanation for members like that, every time.

The only user I see for this is set get_available_bytes.  In that method, I see that we consider the portions of "unavailable" segments past the write pointer to be available.

mark_closed() calls mark_unavail() if we are still initting (!init_complete).  set_open() calls mark_unavail() unconditionally.

I don't understand what the intention here is.  The only case where space past the write_pointer should be considered "available" is if the segment is still open.  I think SegmentCleaner as a whole should simply track the set of open Segments along with their written_to values -- the SegmentProvider interface gets notified when segments are opened, closed, or written to.  written_to doesn't seem to make sense otherwise -- it's irrelevant for a closed segment and 0 for an empty one.

xxhdx1985126 replied:
I changed to track the available space of each segment manager using a counter, please take a look again.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -44,6 +95,249 @@ struct segment_info_t {
+  template <typename T = SegmentInfoT,
+	    std::enable_if_t<std::is_default_constructible<T>::value, int> = 0>
+  void resize(std::vector<SegmentManager*> sms) {
+    assert(sm_info_vec.empty());
```
This rather implies that you cannot call resize() twice.  Let's rename this to initialize_segment_managers()?

athanatos replied:
Do we actually need this at all?  Why not just call add_segment_manager directly?

xxhdx1985126 replied:
> Do we actually need this at all? Why not just call add_segment_manager directly?

yeah, will change that

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -188,52 +499,78 @@ class SpaceTrackerDetailed : public SpaceTrackerI {
+    std::vector<SegmentManager*> sms)
+  {
+    for (auto sm : sms) {
+      if (!sm) {
```
Are there any conditions under which a null pointer here is actually valid?

xxhdx1985126 replied:
yes, ExtentReader::segment_managers may be passed here, and it may contain holes.

athanatos replied:
You need a comment explaining that.  A vector of possibly null pointers is not a normal interface convention.

xxhdx1985126 replied:
will do:-)

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -44,6 +95,249 @@ struct segment_info_t {
+// for keeping track of segment managers' various information,
+// like empty segments, opened segments and so on.
+template <typename SegmentInfoT>
+class segment_info_set_t {
```
Other than the unavailable segment thing, segment_info_set_t makes sense to me now.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -414,40 +746,45 @@ class SegmentCleaner : public SegmentProvider {
     journal_tail_target = journal_seq_t{};
     journal_tail_committed = journal_seq_t{};
     journal_head = journal_seq_t{};
+    journal_device_id = psm.get_device_id();
```
Why not just pass in this device id as journal_device_id?

xxhdx1985126 replied:
there are no device id field in SegmentCleaner

athanatos replied:
SegmentCleaner::journal_device_id.  SegmentCleaner::mount() here accepts a SegmentManager reference, psm, which is only used to obtain an id to populate journal_device_id here.  Just pass in the device_id_t rather than the whole SegmentManager.

xxhdx1985126 replied:
Ah, got it.

athanatos commented on src/crimson/os/seastore/segment_cleaner.h:
```diff
@@ -457,14 +794,19 @@ class SegmentCleaner : public SegmentProvider {
   init_segments_ret init_segments();
 
-  get_segment_ret get_segment() final;
+  get_segment_ret get_segment(device_id_t id = 0) final;
```
Does defaulting to an id of 0 make sense?  Are we guaranteed that there is a device id of 0?

xxhdx1985126 replied:
yes, I think so, I always set "device/block" to device id 0

athanatos replied:
I don't think this makes sense.  The only user I see that doesn't pass an argument is journal.cc:43 in roll_journal_segment.  That usage only makes sense if you guarantee that the journal device always has ID 0.  I see no comment or code that enforces that requirement.

More generally, I don't think we want a caller that forgets to pass an argument to get the 0 device -- we want it to be a compile error.  If we need a call which selects a segment from a device based on a heuristic, we can add that, but it should be explicit.

Remove this default argument and modify that caller in journal.cc to pass the id of the SegmentManager that Journal has a reference to.

xxhdx1985126 replied:
ok, will do

athanatos commented on src/crimson/os/seastore/segment_manager/block.h:
```diff
@@ -165,12 +139,29 @@ class BlockSegmentManager final : public SegmentManager {
+  device_id_t get_device_id() const final {
+    return superblock.device_id;
+  }
+  secondary_device_set_t& get_secondary_devices() final {
```
This shouldn't be part of the SegmentManager interface.  There's a call to get the superblock, that should suffice.

athanatos replied:
Oh, there actually isn't.  Hmm, ok, we can leave this like this for now.

athanatos commented on src/crimson/os/seastore/segment_manager.h:
```diff
@@ -16,6 +17,60 @@
+using magic_t = uint64_t;
+
+using device_spec_t =
+  std::tuple<magic_t, device_type_t, device_id_t>;
```
Eh... let's make this a proper struct with member names and an encode/decode with versioning.  It's ondisk configuration state in a cold path, so we may want to expand this in the future and we don't care about encode/decode performance.

xxhdx1985126 replied:
will do:-)

xxhdx1985126 commented:
> @xuxuehan Once you've addressed the comments I've left so far, I'd like to merge the first three commits on their own and then continue reviewing the rest of the commits. The segment_id_t and ExtentReader changes will probably be ready to merge after those comments are addressed and are likely to cause conflicts with other ongoing work, so merging them quickly is worthwhile.
> 
> If that seems reasonable, I'll continue reviewing the other commits in this PR while you push the first 3 to a new one.

Um, the first four commits were actually splitted from a single commit, I did this just for readability, merging the first three can lead to compile error. Sorry for this inconvenience.

athanatos commented:
> > @xuxuehan Once you've addressed the comments I've left so far, I'd like to merge the first three commits on their own and then continue reviewing the rest of the commits. The segment_id_t and ExtentReader changes will probably be ready to merge after those comments are addressed and are likely to cause conflicts with other ongoing work, so merging them quickly is worthwhile.
> > If that seems reasonable, I'll continue reviewing the other commits in this PR while you push the first 3 to a new one.
> 
> Um, the first four commits were actually splitted from a single commit, I did this just for readability, merging the first three can lead to compilation error, as for the first three commits, the segment cleaner is still referencing old segment_id_t. Sorry for this inconvenience.

Hmm, ok.  I suppose we'll have to merge it all at once then.

xxhdx1985126 commented:
> > > @xuxuehan Once you've addressed the comments I've left so far, I'd like to merge the first three commits on their own and then continue reviewing the rest of the commits. The segment_id_t and ExtentReader changes will probably be ready to merge after those comments are addressed and are likely to cause conflicts with other ongoing work, so merging them quickly is worthwhile.
> > > If that seems reasonable, I'll continue reviewing the other commits in this PR while you push the first 3 to a new one.
> > 
> > 
> > Um, the first four commits were actually splitted from a single commit, I did this just for readability, merging the first three can lead to compilation error, as for the first three commits, the segment cleaner is still referencing old segment_id_t. Sorry for this inconvenience.
> 
> Hmm, ok. I suppose we'll have to merge it all at once then.

Hi, sam, I've addressed the above issues as suggested, please take a look again. Thanks:-)

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,132 @@ struct seastore_meta_t {
+// segment ids without a device id encapsulated
+using device_segment_id_t = uint32_t;
+
+struct segment_id_le_t;
```
I don't think you actually need this for the friend struct segment_id_le_t bit below.

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,132 @@ struct seastore_meta_t {
+    : segment(add_device_id(segment, id)) {
+    // only lower 4 bits are effective, and we have to reserve 0x0F for
+    // special XXX_SEG_IDs
+    assert(id < 15);
```
Let's use the constant defined above?

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,132 @@ struct seastore_meta_t {
+// order of device_id_t
+constexpr uint16_t device_id_len = 4;
+// maximum devices supported
+constexpr uint16_t max_devices = 1 << device_id_len;
```
DEVICE_ID_MAX

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,132 @@ struct seastore_meta_t {
+  static constexpr internal_segment_id_t DEFAULT_INTERNAL_SEG_ID =
+    std::numeric_limits<internal_segment_id_t>::max() - 1;
+public:
+  internal_segment_id_t segment = DEFAULT_INTERNAL_SEG_ID;
```
Make this private

athanatos commented on src/crimson/os/seastore/seastore_types.h:
```diff
@@ -36,28 +36,132 @@ struct seastore_meta_t {
+
+  segment_id_t() = default;
+  constexpr segment_id_t(device_segment_id_t segment)
+    : segment(segment) {}
```
This doesn't really make sense, we want to require callers to explicitly provide a device.

athanatos replied:
See https://github.com/athanatos/ceph/tree/sjust/wip-epm-device

athanatos commented:
https://github.com/athanatos/ceph/tree/sjust/wip-epm-device has some fixes for the comments I made above in seastore_types.h as well as some clarifications for segment_info_set_t::iterator

xxhdx1985126 commented:
@athanatos Thanks, sam! I've cherry-pick your commits and repushed this pr, please take a look again:-)

athanatos commented:
@xxhdx1985126 Ok, I've repushed it

xxhdx1985126 commented:
@athanatos Thanks:-)

athanatos commented:
I'm still working on simplifying segment_info_set_t, should have something new pushed tomorrow.
